File: cli/src/voidrift_cli/manifest.py
"""Task manifest manager — CLI-owned orchestration state (REQ-TM-1 through REQ-TM-7).

Reads/writes .voidrift/tasks/manifest.yml. Tracks task status, dependencies,
module grouping, bug references. No agent tool provides write access.
"""

from __future__ import annotations

import shutil
from datetime import datetime
from pathlib import Path

import yaml


class ManifestManager:
    """Manages .voidrift/tasks/manifest.yml and task file lifecycle."""

    def __init__(self, project_dir: Path | None = None) -> None:
        self._project = (project_dir or Path.cwd()).resolve()
        self._tasks_dir = self._project / ".voidrift" / "tasks"
        self._manifest_path = self._tasks_dir / "manifest.yml"
        self._active_dir = self._tasks_dir / "active"
        self._archived_dir = self._tasks_dir / "archived"
        self._history_path = self._tasks_dir / "history.log"
        self._data: dict = {}

    # ── Bootstrap ────────────────────────────────────────────────────────

    def ensure_dirs(self) -> None:
        """Create task directories if they don't exist."""
        self._active_dir.mkdir(parents=True, exist_ok=True)
        self._archived_dir.mkdir(parents=True, exist_ok=True)

    def exists(self) -> bool:
        """Return True if manifest.yml exists."""
        return self._manifest_path.exists()

    # ── Load / Save ──────────────────────────────────────────────────────

    def load(self) -> None:
        """Load manifest from disk."""
        if self._manifest_path.exists():
            self._data = yaml.safe_load(self._manifest_path.read_text()) or {}
        else:
            self._data = {}

    def save(self) -> None:
        """Write manifest to disk."""
        self._tasks_dir.mkdir(parents=True, exist_ok=True)
        self._manifest_path.write_text(yaml.dump(self._data, default_flow_style=False, sort_keys=False))

    # ── Accessors ────────────────────────────────────────────────────────

    @property
    def next_id(self) -> int:
        return self._data.get("next_id", 1)

    @property
    def next_bug_id(self) -> int:
        return self._data.get("next_bug_id", 1)

    def tasks(self) -> dict[int, dict]:
        """Return all active tasks: {id: {status, module, ...}}."""
        return self._data.get("tasks", {})

    def modules(self) -> dict[str, list[int]]:
        """Return module groupings: {name: [task_ids]}."""
        return self._data.get("modules", {})

    def dependencies(self) -> dict[int, list[int]]:
        """Return dependency graph: {task_id: [depends_on_ids]}."""
        return self._data.get("dependencies", {})

    def get_task(self, task_id: int) -> dict | None:
        """Return a single task's manifest entry."""
        return self.tasks().get(task_id)

    def get_status(self, task_id: int) -> str:
        """Return a task's status, or empty string if not found."""
        t = self.get_task(task_id)
        return t.get("status", "") if t else ""

    # ── Dispatch ─────────────────────────────────────────────────────────

    def dispatchable(self) -> list[int]:
        """Return task IDs that are ready to dispatch (REQ-TM-3).

        A task is dispatchable when status=planned and all dependencies
        are implemented or verified.
        """
        tasks = self.tasks()
        deps = self.dependencies()
        ready = []
        for tid, t in tasks.items():
            if t.get("status") != "planned":
                continue
            task_deps = deps.get(tid, [])
            if all(self.get_status(d) in ("implemented", "verified") for d in task_deps):
                ready.append(tid)
        return ready

    def has_work(self) -> bool:
        """Return True if any tasks are not verified/blocked."""
        return any(t.get("status") in ("planned", "in-progress", "implemented", "failed")
                   for t in self.tasks().values())

    # ── Status Transitions (REQ-TM-2) ───────────────────────────────────

    def set_status(self, task_id: int, status: str) -> None:
        """Set a task's status and handle side effects."""
        tasks = self._data.setdefault("tasks", {})
        if task_id not in tasks:
            return
        tasks[task_id]["status"] = status

        if status == "failed":
            self._block_dependents(task_id)
        elif status in ("implemented", "verified"):
            self._unblock_dependents(task_id)

        self.save()

    def _block_dependents(self, failed_id: int) -> None:
        """Transitively block all tasks depending on failed_id."""
        deps = self.dependencies()
        tasks = self.tasks()
        to_block = set()
        frontier = [tid for tid, d in deps.items() if failed_id in d]
        while frontier:
            tid = frontier.pop()
            if tid in to_block:
                continue
            to_block.add(tid)
            if tasks.get(tid, {}).get("status") in ("planned", "in-progress"):
                tasks[tid]["status"] = "blocked"
            frontier.extend(t for t, d in deps.items() if tid in d)

    def _unblock_dependents(self, resolved_id: int) -> None:
        """Unblock tasks whose dependencies are now all met."""
        deps = self.dependencies()
        tasks = self.tasks()
        for tid, task_deps in deps.items():
            if resolved_id not in task_deps:
                continue
            if tasks.get(tid, {}).get("status") != "blocked":
                continue
            if all(self.get_status(d) in ("implemented", "verified") for d in task_deps):
                tasks[tid]["status"] = "planned"

    # ── Task File Access ─────────────────────────────────────────────────

    def task_path(self, task_id: int) -> Path:
        """Return the path to a task file in active/."""
        return self._active_dir / f"TASK-{task_id}.md"

    def bug_path(self, bug_id: int) -> Path:
        """Return the path to a bug file in active/."""
        return self._active_dir / f"BUG-{bug_id}.md"

    def read_task(self, task_id: int) -> str:
        """Read a task file's content."""
        p = self.task_path(task_id)
        if p.exists():
            return p.read_text()
        return ""

    # ── Registration ─────────────────────────────────────────────────────

    def add_task(self, task_id: int, module: str, depends: list[int] | None = None) -> None:
        """Register a task in the manifest."""
        tasks = self._data.setdefault("tasks", {})
        tasks[task_id] = {"status": "planned", "module": module}
        modules = self._data.setdefault("modules", {})
        modules.setdefault(module, [])
        if task_id not in modules[module]:
            modules[module].append(task_id)
        if depends:
            deps = self._data.setdefault("dependencies", {})
            deps[task_id] = depends
        self._data["next_id"] = max(self.next_id, task_id + 1)
        self.save()

    def add_bug(self, bug_id: int, refs: list[int] | None = None) -> None:
        """Register a bug in the manifest."""
        bugs = self._data.setdefault("bugs", {})
        bugs[bug_id] = {"status": "open", "refs": refs or []}
        self._data["next_bug_id"] = max(self.next_bug_id, bug_id + 1)
        if refs:
            tasks = self.tasks()
            for tid in refs:
                if tid in tasks:
                    tasks[tid].setdefault("refs", [])
                    if bug_id not in tasks[tid]["refs"]:
                        tasks[tid]["refs"].append(bug_id)
        self.save()

    # ── Archive (REQ-TM-5, REQ-TM-6) ────────────────────────────────────

    def archive(self, task_id: int) -> None:
        """Move a verified task to archived/ and log to history."""
        tasks = self._data.get("tasks", {})
        if task_id not in tasks:
            return
        task = tasks[task_id]
        module = task.get("module", "")
        assigned = task.get("assigned", "")

        # Move task file
        src = self.task_path(task_id)
        dst = self._archived_dir / src.name
        if src.exists():
            self._archived_dir.mkdir(parents=True, exist_ok=True)
            shutil.move(str(src), str(dst))

        # Move referenced bug files if all their task refs are verified
        for bug_id in task.get("refs", []):
            bugs = self._data.get("bugs", {})
            bug = bugs.get(bug_id, {})
            bug_refs = bug.get("refs", [])
            if all(self.get_status(r) == "verified" or r == task_id for r in bug_refs):
                bug_src = self.bug_path(bug_id)
                if bug_src.exists():
                    shutil.move(str(bug_src), str(self._archived_dir / bug_src.name))
                bugs.pop(bug_id, None)

        # Remove from manifest
        del tasks[task_id]
        modules = self.modules()
        for mod_tasks in modules.values():
            if task_id in mod_tasks:
                mod_tasks.remove(task_id)
        deps = self._data.get("dependencies", {})
        deps.pop(task_id, None)

        # Append to history log
        ts = datetime.now().isoformat(timespec="seconds")
        with open(self._history_path, "a") as f:
            f.write(f"{ts} TASK-{task_id} verified module={module} assigned={assigned}\n")

        self.save()

    # ── Ideas (REQ-IDEA-1 through REQ-IDEA-4) ──────────────────────────

    @property
    def next_idea_id(self) -> int:
        return self._data.get("next_idea_id", 1)

    def ideas(self) -> dict[int, dict]:
        """Return all ideas: {id: {status}}."""
        return self._data.get("ideas", {})

    def idea_path(self, idea_id: int) -> Path:
        """Return the path to an idea file in active/."""
        return self._active_dir / f"IDEA-{idea_id}.md"

    def add_idea(self, idea_id: int, status: str = "draft") -> None:
        """Register an idea in the manifest."""
        ideas = self._data.setdefault("ideas", {})
        ideas[idea_id] = {"status": status}
        self._data["next_idea_id"] = max(self.next_idea_id, idea_id + 1)
        self.save()

    def set_idea_status(self, idea_id: int, status: str) -> None:
        """Update an idea's status."""
        ideas = self._data.get("ideas", {})
        if idea_id in ideas:
            ideas[idea_id]["status"] = status
            self.save()

    def read_idea(self, idea_id: int) -> str:
        """Read an idea file's content."""
        p = self.idea_path(idea_id)
        return p.read_text() if p.exists() else ""

    # ── Summary ──────────────────────────────────────────────────────────

    def summary(self) -> dict[str, int]:
        """Return counts by status."""
        counts: dict[str, int] = {}
        for t in self.tasks().values():
            s = t.get("status", "unknown")
            counts[s] = counts.get(s, 0) + 1
        return counts

File: src/manifest.ts
/**
 * ManifestManager: task status, deps, dispatch (REQ-TM-1..7).
 */

import { readFileSync, writeFileSync, existsSync, mkdirSync, renameSync, appendFileSync } from "node:fs";
import { join } from "node:path";
import { parse as parseYaml, stringify as stringifyYaml } from "yaml";

export type TaskStatus = "planned" | "in-progress" | "implemented" | "verified" | "failed" | "blocked";

export interface TaskEntry {
  status: TaskStatus;
  module: string;
  depends: number[];
  skills: string[];
  reqs: string[];
  title?: string;
  idea?: number;
}

export class ManifestManager {
  private _dir: string;
  private _path: string;
  private _tasks: Record<number, TaskEntry> = {};
  private _nextId = 1;
  private _nextBugId = 1;

  constructor(projectDir?: string) {
    this._dir = join(projectDir ?? process.cwd(), ".voidrift");
    this._path = join(this._dir, "tasks", "manifest.yml");
  }

  exists(): boolean { return existsSync(this._path); }

  load(): void {
    if (!this.exists()) return;
    const raw = parseYaml(readFileSync(this._path, "utf-8")) ?? {};
    this._nextId = raw.next_id ?? 1;
    this._nextBugId = raw.next_bug_id ?? 1;
    this._tasks = {};
    for (const [id, entry] of Object.entries(raw.tasks ?? {})) {
      const e = entry as Record<string, unknown>;
      this._tasks[Number(id)] = {
        status: (e.status as TaskStatus) ?? "planned",
        module: String(e.module ?? ""),
        depends: Array.isArray(e.depends) ? e.depends.map(Number) : [],
        skills: Array.isArray(e.skills) ? e.skills.map(String) : [],
        reqs: Array.isArray(e.reqs) ? e.reqs.map(String) : [],
        title: e.title ? String(e.title) : undefined,
        idea: e.idea != null ? Number(e.idea) : undefined,
      };
    }
  }

  save(): void {
    mkdirSync(join(this._dir, "tasks"), { recursive: true });
    const data: Record<string, unknown> = {
      next_id: this._nextId,
      next_bug_id: this._nextBugId,
      tasks: {} as Record<string, unknown>,
    };
    for (const [id, t] of Object.entries(this._tasks)) {
      (data.tasks as Record<string, unknown>)[id] = { ...t };
    }
    writeFileSync(this._path, stringifyYaml(data), "utf-8");
  }

  tasks(): Record<number, TaskEntry> { return { ...this._tasks }; }

  getTask(id: number): TaskEntry | null { return this._tasks[id] ?? null; }

  setStatus(id: number, status: TaskStatus): void {
    const task = this._tasks[id];
    if (!task) return;
    task.status = status;
    this.save();
    // Transitive blocking (REQ-TM-3)
    if (status === "failed") {
      for (const [tid, t] of Object.entries(this._tasks)) {
        if (t.depends.includes(id) && t.status === "planned") {
          t.status = "blocked";
        }
      }
      this.save();
    }
    // Unblock dependents when implemented
    if (status === "implemented" || status === "verified") {
      for (const [tid, t] of Object.entries(this._tasks)) {
        if (t.status === "blocked" && t.depends.includes(id)) {
          const allMet = t.depends.every(d => {
            const dep = this._tasks[d];
            return dep && (dep.status === "implemented" || dep.status === "verified");
          });
          if (allMet) t.status = "planned";
        }
      }
      this.save();
    }
    this._appendHistory(id, status);
    // Auto-archive idea when all its tasks are verified (REQ-P-8)
    if (status === "verified" && task.idea) this.checkIdeaArchival(task.idea);
  }

  hasWork(): boolean {
    return Object.values(this._tasks).some(t =>
      t.status === "planned" || t.status === "in-progress"
    );
  }

  dispatchable(): number[] {
    return Object.entries(this._tasks)
      .filter(([, t]) => {
        if (t.status !== "planned") return false;
        return t.depends.every(d => {
          const dep = this._tasks[d];
          return dep && (dep.status === "implemented" || dep.status === "verified");
        });
      })
      .map(([id]) => Number(id));
  }

  summary(): Record<TaskStatus, number> {
    const counts: Record<TaskStatus, number> = {
      planned: 0, "in-progress": 0, implemented: 0, verified: 0, failed: 0, blocked: 0,
    };
    for (const t of Object.values(this._tasks)) counts[t.status]++;
    return counts;
  }

  readIdea(id: number): string | null {
    const p = join(this._dir, "ideas", `IDEA-${id}.md`);
    return existsSync(p) ? readFileSync(p, "utf-8") : null;
  }

  ideaPath(id: number): string {
    return join(this._dir, "ideas", `IDEA-${id}.md`);
  }

  archive(id: number): void {
    const activeDir = join(this._dir, "tasks", "active");
    const archiveDir = join(this._dir, "tasks", "archived");
    mkdirSync(archiveDir, { recursive: true });
    const src = join(activeDir, `TASK-${id}.md`);
    if (existsSync(src)) renameSync(src, join(archiveDir, `TASK-${id}.md`));
    delete this._tasks[id];
    this.save();
  }

  /** Create a bug entity (REQ-TM-7). Returns bug ID. */
  createBug(title: string, content: string, taskRefs: number[] = []): number {
    const bugId = this._nextBugId++;
    const bugDir = join(this._dir, "tasks", "active");
    mkdirSync(bugDir, { recursive: true });
    const frontmatter = `---\nid: BUG-${bugId}\ntitle: ${title}\ntask_refs: [${taskRefs.join(", ")}]\n---\n\n`;
    writeFileSync(join(bugDir, `BUG-${bugId}.md`), frontmatter + content, "utf-8");
    this.save();
    this._appendHistory(bugId, "created" as TaskStatus);
    return bugId;
  }

  /** Check if all tasks for an idea are verified and archive the idea (REQ-IDEA-5). */
  checkIdeaArchival(ideaId: number): boolean {
    const ideaTasks = Object.values(this._tasks).filter(t => t.idea === ideaId);
    if (!ideaTasks.length) return false;
    if (ideaTasks.every(t => t.status === "verified")) {
      // Archive the idea
      const ideaPath = join(this._dir, "ideas", `IDEA-${ideaId}.md`);
      const archiveDir = join(this._dir, "ideas", "archived");
      mkdirSync(archiveDir, { recursive: true });
      if (existsSync(ideaPath)) {
        renameSync(ideaPath, join(archiveDir, `IDEA-${ideaId}.md`));
        this._appendHistory(ideaId, "done" as TaskStatus);
        return true;
      }
    }
    return false;
  }

  private _appendHistory(id: number, status: TaskStatus): void {
    const historyPath = join(this._dir, "tasks", "history.log");
    const task = this._tasks[id];
    const ts = new Date().toISOString();
    const line = `${ts} TASK-${id} ${status} module=${task?.module ?? ""}\n`;
    try { appendFileSync(historyPath, line); } catch { /* */ }
  }
}
